whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java

```diff
@@ -37,46 +38,68 @@ public MessageService(
+        if (isNotValidMessage(message, outgoingEmail)){
             return;
         }
 
-        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        message.emailAddresses().forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE));
     }
 
-    public void verifyAndSetAsUnsubscribed(final MessageInfo message){
-        final String email = outgoingMessageDao.getEmail(message.messageId());
+    public void verifyAndSetAsUnsubscribed(final EmailMessageInfo message){
+        if (message.emailAddresses() != null && message.emailAddresses().size() > 1){
+            LOGGER.error("This can not happen, unsubscribe with multiple recipients");
+            return;
+        }
+
+        final List<String> outgoingEmail = outgoingMessageDao.getEmails(message.messageId());
 
-        if (isIncorrectMessage(message, email)){
+        if (isNotValidMessage(message, outgoingEmail)){
             return;
         }
 
-        LOGGER.debug("Unsubscribe message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNSUBSCRIBE);
+        LOGGER.debug("Unsubscribe message-id {} email {}", message.messageId(), message.emailAddresses().get(0));
+        emailStatusDao.createEmailStatus(message.emailAddresses().get(0), EmailStatus.UNSUBSCRIBE);
     }
 
-    private boolean isIncorrectMessage(final MessageInfo message, final String email){
-        if (Strings.isNullOrEmpty(email)) {
+    private boolean isNotValidMessage(final EmailMessageInfo message, final List<String> outgoingEmail){
+        if (message.messageId() == null || message.emailAddresses() == null || message.emailAddresses().isEmpty()){
+            LOGGER.warn("Incorrect message {}", message.messageId());
+            return true;
+        }
+
+        if (outgoingEmail == null || outgoingEmail.isEmpty()) {
             LOGGER.warn("Couldn't find outgoing message matching {}", message.messageId());
             return true;
         }
 
-        if (!email.equalsIgnoreCase(message.emailAddress())) {
-            LOGGER.warn("Email {} in outgoing message doesn't match '{}' in failure response", email, message.emailAddress());
+        if (!containsAllCaseInsensitive(message.emailAddresses(), outgoingEmail)) {
+            LOGGER.warn("Email {} in outgoing message doesn't match '{}' in failure response", outgoingEmail, StringUtils.join(message.emailAddresses(), ", "));
             return true;
         }
         return false;
     }
 
+    private boolean containsAllCaseInsensitive(final List<String> messageRecipients, final List<String> storedEmails){
+        final List<String> emailsInLowerCase = storedEmails
```

Comment:
It will always be false specially incase of unsubscribe? It should be contains rather then matchAll?

Reply:
messageRecipients -> test@ripe.net, test1@ripe.net
emailsInLowerCase (email in outgoing table) -> test@ripe.net, test1@ripe.net, test2@ripe.net

test2@ripe.net -> succeeded

loop:
(test@ripe.net, test1@ripe.net).allMatch((test@ripe.net, test1@ripe.net, test2@ripe.net).contains(test@ripe.net))
(test@ripe.net, test1@ripe.net).allMatch((test@ripe.net, test1@ripe.net, test2@ripe.net).contains(test1@ripe.net))

It is fine, right? In case if unsubscribe just modify messageRecipients

Reply:
and the tests are green with this